addons/mail/tests/test_mail_canned_response.py

```diff
@@ -0,0 +1,78 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.addons.mail.tests.common import MailCommon
+from odoo.exceptions import AccessError
+from odoo.tests.common import tagged, HttpCase
+from odoo import Command
+
+
+@tagged("mail_canned_response", "post_install", "-at_install")
+class TestCannedResponse(MailCommon, HttpCase):
+
+    def test_user_canned_response(self):
+        self.authenticate(self.user_employee.login, self.user_employee.login)
+        self.env["mail.canned.response"].with_user(self.user_employee).create({
+            "source": "hlp",
+            "substitution": "help"
+        })
+        request = self.make_jsonrpc_request("/mail/data", {"canned_responses": True})
+        self.assertEqual(len(request["CannedResponse"]), 1)
+
+    def test_user_group_canned_response(self):
+        self.authenticate(self.user_employee.login, self.user_employee.login)
+        self.env["mail.canned.response"].with_user(self.user_admin).create({
+            "source": "hlp",
+            "substitution": "help",
+            "group_ids": [Command.set([self.ref("base.group_user")])],
+        })
+        request = self.make_jsonrpc_request("/mail/data", {"canned_responses": True})
+        self.assertEqual(len(request["CannedResponse"]), 1)
+
+    def test_user_no_group_canned_response(self):
+        self.authenticate(self.user_employee.login, self.user_employee.login)
+        self.env["mail.canned.response"].with_user(self.user_admin).create({
+            "source": "hlp",
+            "substitution": "help",
+            'group_ids': [Command.set([self.ref('base.group_system')])],
+        })
+        request = self.make_jsonrpc_request("/mail/data", {"canned_responses": True})
+        self.assertEqual(request, {})
+
+    def test_user_can_write_own_canned_response(self):
+        self.authenticate(self.user_employee.login, self.user_employee.login)
+        canned_response = self.env["mail.canned.response"].with_user(self.user_employee).create({
+            "source": "hlp",
+            "substitution": "help",
+        })
+        canned_response.with_user(self.user_employee).write({"source": "help"})
+        self.assertEqual(canned_response.source, "help")
+
+    def test_creator_write_shared_canned_response(self):
+        self.authenticate(self.user_employee.login, self.user_employee.login)
+        canned_response = self.env["mail.canned.response"].with_user(self.user_employee).create({
```

Comment:
This should be done in a setUpClass as it is redundant code that does not belong to the test itself, more to data preparation (as it is always the same it seems, globally). You may create a global_response and user_response in setUp to ease test writing and understanding.